File: annotations-runtime/annotations-runtime-50/src/main/java/com/github/dynamicextensionsalfresco/quartz/QuartzJobRegistrar.java
package com.github.dynamicextensionsalfresco.quartz;

import com.github.dynamicextensionsalfresco.jobs.ScheduledQuartzJob;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Map;
import java.util.Properties;
import org.alfresco.repo.lock.JobLockService;
import org.alfresco.schedule.AbstractScheduledLockedJob;
import org.quartz.CronTrigger;
import org.quartz.Job;
import org.quartz.JobDataMap;
import org.quartz.JobDetail;
import org.quartz.Scheduler;
import org.quartz.SchedulerException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.util.Assert;

/**
 * Created by jasper on 19/07/17.
 */
public class QuartzJobRegistrar implements ApplicationContextAware, InitializingBean, DisposableBean {
    private Logger logger = LoggerFactory.getLogger(QuartzJobRegistrar.class);

    public final static String BEAN_ID="bean";
    public static final String JOB_LOCK_SERVICE = "jobLockService";

    private ArrayList<ScheduledQuartzJob> registeredJobs = new ArrayList<ScheduledQuartzJob>();

    @Autowired
    protected Scheduler scheduler;

    @Autowired
    protected JobLockService jobLockService;

    @Autowired
    @Qualifier("global-properties")
    protected Properties globalProperties = new Properties();

    private ApplicationContext applicationContext;

    @Override
    public void afterPropertiesSet() {
        this.registerScheduledQuartzJobAnnotatedBeans();
    }

    public void registerScheduledQuartzJobAnnotatedBeans() {
        Map<String, Object> scheduledBeans = applicationContext.getBeansWithAnnotation(ScheduledQuartzJob.class);
        for (Map.Entry entry : scheduledBeans.entrySet()) {
            Object bean = entry.getValue();
            try {
                this.registerJob(bean);
            } catch (Exception e) {
                logger.error("Failed to register job: ", e);
            }
        }
    }

    public void registerJob(Object bean) throws ParseException, SchedulerException {
        Assert.isInstanceOf(Job.class, bean, "annotated Quartz job classes should implement org.quartz.Job");

        ScheduledQuartzJob annotation = bean.getClass().getAnnotation(ScheduledQuartzJob.class);

        String cron = applicationContext.getBean("global-properties", Properties.class).getProperty(annotation.cronProp(), annotation.cron());
        CronTrigger trigger = new CronTrigger(annotation.name(), annotation.group(), cron);
        JobDetail jobDetail = new JobDetail(annotation.name(), annotation.group(), annotation.cluster() ? ClusterLockedJob.class : GenericQuartzJob.class);

        JobDataMap map = new JobDataMap();
        map.put(BEAN_ID, bean);
        map.put(JOB_LOCK_SERVICE, this.jobLockService);

        jobDetail.setJobDataMap(map);
        scheduler.scheduleJob(jobDetail, trigger);

        registeredJobs.add(annotation);

        logger.debug("scheduled job " + annotation.name() + " from group " + annotation.group() + " using cron " + annotation.cron());
    }

    @Override
    public void destroy() {
        for (ScheduledQuartzJob job : registeredJobs) {
            try {
                scheduler.unscheduleJob(job.name(), job.group());
                logger.debug("unscheduled job " + job.name() + " from group " + job.group());
            } catch (SchedulerException e) {
                logger.error("failed to cleanup quartz job " + job, e);
            }
        }
    }

    public void setApplicationContext(ApplicationContext applicationContext) {
        this.applicationContext = applicationContext;
    }

    public void setScheduler(Scheduler scheduler) {
        this.scheduler = scheduler;
    }

    public void setJobLockService(JobLockService jobLockService) {
        this.jobLockService = jobLockService;
    }

}

File: annotations-runtime/src/main/kotlin/com/github/dynamicextensionsalfresco/quartz/QuartzJobRegistrar.kt
package com.github.dynamicextensionsalfresco.quartz

import com.github.dynamicextensionsalfresco.debug
import com.github.dynamicextensionsalfresco.jobs.ScheduledQuartzJob
import org.quartz.*
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.DisposableBean
import org.springframework.beans.factory.InitializingBean
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware
import org.springframework.util.Assert
import java.util.*

/**
 * Look for beans annotated with @ScheduledJob and register/unregister them with Quartz
 *
 * @author Laurent Van der Linden
 */
public class QuartzJobRegistrar @Autowired constructor(var scheduler: Scheduler) : ApplicationContextAware, InitializingBean, DisposableBean {
    private val logger = LoggerFactory.getLogger(QuartzJobRegistrar::class.java)

    private val registeredJobs = ArrayList<ScheduledQuartzJob>()

    private var applicationContext: ApplicationContext? = null

    @Throws(Exception::class)
    override fun afterPropertiesSet() {
        val scheduledBeans = applicationContext!!.getBeansWithAnnotation(ScheduledQuartzJob::class.java)
        for (entry in scheduledBeans.entries) {
            val bean = entry.value

            Assert.isInstanceOf(Job::class.java, bean, "annotated Quartz job classes should implement org.quartz.Job")

            val annotation = bean.javaClass.getAnnotation(ScheduledQuartzJob::class.java)

            try {
                val cron = applicationContext!!.getBean("global-properties", Properties::class.java).getProperty(annotation.cronProp, annotation.cron);
                val trigger = CronTrigger(annotation.name, annotation.group, cron)
                val jobDetail = JobDetail(annotation.name, annotation.group, GenericQuartzJob::class.java)
                jobDetail.jobDataMap.put(GenericQuartzJob.BEAN_ID, bean)
                scheduler.scheduleJob(jobDetail, trigger)

                registeredJobs.add(annotation)

                logger.debug { "scheduled job ${annotation.name} from group ${annotation.group} using cron ${annotation.cron}" }
            } catch (e: Exception) {
                logger.error("failed to register job ${annotation.name} using cron ${annotation.cron}", e)
            }

        }
    }

    @Throws(Exception::class)
    override fun destroy() {
        for (job in registeredJobs) {
            try {
                scheduler.unscheduleJob(job.name, job.group)
                logger.debug { "unscheduled job ${job.name} from group ${job.group}" }
            } catch (e: SchedulerException) {
                logger.error("failed to cleanup quartz job " + job, e)
            }
        }
    }

    override fun setApplicationContext(applicationContext: ApplicationContext?) {
        this.applicationContext = applicationContext
    }
}
